quivr/linkage.py

```diff
@@ -0,0 +1,164 @@
+
+        if len(right_keys) != len(right_table):
+            raise ValueError("Right keys must have the same length as the right table")
+
+        self.left_table = left_table
+        self.right_table = right_table
+
+        self.left_index = ArrowArrayIndex(left_keys)
+        self.right_index = ArrowArrayIndex(right_keys)
+
+        self.all_unique_values = self.left_index.values.union(self.right_index.values)
+
+    def select_left(self, val: Any) -> tables.Table:
+        """
+        Select the rows from the left table that match the given value.
+
+        If the value is not present in the left table, then an empty table is
+        returned.
+        """
+        if not isinstance(val, pa.Scalar):
+            val = pa.scalar(val)
+        return self._select_left(val)
+
+    def _select_left(self, val: pa.Scalar) -> tables.Table:
+        left_indices = self.left_index.get(val)
+        if left_indices is None:
+            return self.left_table.empty()
+        else:
+            return self.left_table.take(left_indices)
+
+    def select_right(self, val: Any) -> tables.Table:
+        """
+        Select the rows from the right table that match the given value.
+
+        If the value is not present in the right table, then an empty table is
+        returned.
+        """
+        if not isinstance(val, pa.Scalar):
+            val = pa.scalar(val)
+        return self._select_right(val)
+
+    def _select_right(self, val: pa.Scalar) -> tables.Table:
+        right_indices = self.right_index.get(val)
+        if right_indices is None:
+            return self.right_table.empty()
+        else:
+            return self.right_table.take(right_indices)
+
+    def select(self, val: Any) -> tuple[tables.Table, tables.Table]:
+        """
+        Select the rows from both tables that match the given value.
+
+        If the value is not present in either table, then an empty table is
+        returned for that table.
+        """
+        if not isinstance(val, pa.Scalar):
+            val = pa.scalar(val)
+        return self._select_left(val), self._select_right(val)
+
+    def __getitem__(self, val: Any):
+        return self.select(val)
+
+    def iterate(self) -> Iterator[tuple[pa.Scalar, tables.Table, tables.Table]]:
+        """
+        Returns an iterator over all the unique values in the linkage, and the rows from
+        each table that match that value.
+        """
+        for val in self.all_unique_values:
+            yield val, self._select_left(val), self._select_right(val)
+
+    def __iter__(self):
+        return self.iterate()
+
+    def __len__(self):
+        """Returns the number of unique values in the linkage."""
+        return len(self.all_unique_values)
+
+
+def composite_array(*keys: list[pa.Array]) -> pa.Array:
+    """
+    Create a composite array from a list of arrays.
+    """
+    fields = [pa.field(f"key_{i}", k.type) for i, k in enumerate(keys)]
```

Comment:
For clarity, having the names of the keys be the names of the array would be helpful.